### PR TITLE
BUG: Fix crash on 0d return value in apply_along_axis

### PR DESCRIPTION
**Pull request by eric-wieser**
Also:
ENH: Support arbitrary dimensionality of return value
MAINT: remove special casing

Now supports
```python
>>> data = np.arange(6).reshape(2, 3)
>>> data
array([[0, 1, 2],
       [3, 4, 5]])
>>> np.apply_along_axis(np.diag, -1, data)  #would previously crash
array([[[0, 0, 0],
        [0, 1, 0],
        [0, 0, 2]],

       [[3, 0, 0],
        [0, 4, 0],
        [0, 0, 5]]])
>>> np.apply_along_axis(lambda x: np.array(np.sum(x)), 0, data)  #would previously crash
array([3, 5, 7])
```

Addresses [discussion from #8363](https://github.com/numpy/numpy/pull/8363#issuecomment-270027102)

### CODE REVIEW

File: numpy/lib/shape_base.py [hunk 1]
```diff
@@ -74,65 +75,53 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
+
+    # outarr, with inserted dimensions at the end
+    # this means that outarr_view[i] = func1d(inarr_view[i])
+    dims_out = list(range(outarr.ndim))
+    outarr_view = np.transpose(outarr, dims_out[:axis] + dims_out[axis+res.ndim:] + dims_out[axis:axis+res.ndim])
+
+    # save the first result
+    outarr_view[tuple(ind)] = res
+    k = 1
+    while k < Ntot:
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Is this really likely to be faster than using `ndindex` here?

File: numpy/lib/shape_base.py [hunk 2]
```diff
@@ -74,65 +76,42 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
-        if res.shape == ():
-            outarr = outarr.squeeze(axis)
-        return outarr
+
+    # arr, with the iteration axis at the end
+    dims_in = list(range(nd))
+    inarr_view = transpose(arr, dims_in[:axis] + dims_in[axis+1:] + [axis])
+
+    # indices
+    inds = ndindex(inarr_view.shape[:nd-1])
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: To be honest, most of this function could probably be replaced with a carefully constructed `nditer` that does the transposes for us, but this is probably good enough for now

**Comment by eric-wieser** [comment]
Right now, this happens:
```
# silent broadcasting
>>> apply_along_axis(lambda x: 1 if x.all() else [2, 3], 0, np.array([[False, True, True]]))))
array([[2, 1, 1],
       [3, 1, 1]])
# same function, different order
>>> apply_along_axis(lambda x: 1 if x.all() else [2, 3], 0, np.array([[True, False, True]]))))
ValueError: setting an array element with a sequence.
```

Would it be better if both raised `ValueError: shape must be the same on each invocation`, at the expense of a check on every iteration?

**Comment by eric-wieser** [comment]
#5875 is rearing its head in the errors here...

**Comment by eric-wieser** [comment]
This fails when `f()` returns a `matrix`, because `matrix`s violate the assumption that `arr[0].shape == arr.shape[1:]`

File: numpy/lib/shape_base.py [hunk 3]
```diff
@@ -100,7 +101,11 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
     # insert as many axes as necessary to create the output
     outshape = arr.shape[:axis] + res.shape + arr.shape[axis+1:]
     outarr = zeros(outshape, res.dtype)
-    outarr = res.__array_prepare__(outarr)
+    # matrices call reshape in __array_prepare__, and this is apparently ok!
+    if not isinstance(res, matrix):
+        outarr = res.__array_prepare__(outarr)
+    if outshape != outarr.shape:
+        raise ValueError('__array_prepare__ should not change the shape of the resultant array')
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Is this assumption supposed to be valid? How do we deal with `matrix` violating it elsewhere?

File: numpy/lib/tests/test_shape_base.py [hunk 4]
```diff
@@ -28,13 +28,23 @@ def test_3d(self):
...
+
         result = apply_along_axis(double, 0, m)
         assert isinstance(result, np.matrix)
         assert_array_equal(
-            result, np.matrix([[0, 2], [4, 6]])
+            result, expected
+        )
+
+        result = apply_along_axis(double, 1, m)
+        assert isinstance(result, np.matrix)
```
Comment (rgommers) [inline numpy/lib/tests/test_shape_base.py]: Note that assert shouldn't be used at all, use np.testing.assert_ instead (applies to multiple lines below as well)

Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: In my defense, those other lines were not part of this patch. Do you want me to fix all cases of assert in this file anyway?

Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: Or should this be `self.assertIsInstance`?

Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: Changed to `assert_(isinstance(...))`, because that's what other tests do. Switching to `self.assertIsInstance` can be discussed elsewhere

Comment (rgommers) [inline numpy/lib/tests/test_shape_base.py]: > In my defense, those other lines were not part of this patch. Do you want me to fix all cases of assert in this file anyway?

if you could, that'd be useful

Comment (rgommers) [inline numpy/lib/tests/test_shape_base.py]: ``self.assertIsInstance`` seems fine to use if you prefer that, it has a sane implementation in both ``unittest`` and ``unittest2``.

Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: @rgommers: I've patched the other ones in that file

**Comment by eric-wieser** [comment]
Ugh, this needs a patch in [`ma.apply_along_axis`](https://github.com/numpy/numpy/blob/v1.12.0/numpy/ma/extras.py#L377-L461) as well.

File: numpy/lib/shape_base.py [hunk 5]
```diff
@@ -73,66 +76,66 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
+    ind0 = next(inds)
+    res = func1d(inarr_view[ind0], *args, **kwargs)
+    res = asanyarray(res)
+
+    # insert as many axes as necessary to create the output
+    outshape = arr.shape[:axis] + res.shape + arr.shape[axis+1:]
+    outarr = zeros(outshape, res.dtype)
+    # matrices call reshape in __array_prepare__, and this is apparently ok!
+    if not isinstance(res, matrix):
+        outarr = res.__array_prepare__(outarr)
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Can someone comment on whether this is actually correct, or if it can just be omitted altogether

`__array_prepare__` seems to always do something screwy for builtin array subclasses. It's not clear to me what the contract entails, and whether I'm violating it, or `np.matrix` and `np.ma.MaskedArray` are

Comment (shoyer) [inline numpy/lib/shape_base.py]: @mhvk may know

File: numpy/lib/shape_base.py [hunk 6]
```diff
@@ -73,66 +76,66 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
+
+    # outarr, with inserted dimensions at the end
+    # this means that outarr_view[i] = func1d(inarr_view[i])
+    dims_out = list(range(outarr.ndim))
+    outarr_view = transpose(outarr, dims_out[:axis] + dims_out[axis+res.ndim:] + dims_out[axis:axis+res.ndim])
+
+    # save the first result
+    outarr_view[ind0] = res
+    for ind in inds:
+        outarr_view[ind] = asanyarray(func1d(inarr_view[ind], *args, **kwargs))
```
Comment (shoyer) [inline numpy/lib/shape_base.py]: I recognize that this works to update the view inplace, but it still makes me a little nervous. It would be more direct and obviously correct to do the transposing afterwards.

This would change the memory layout of the result array. Transposing afterwards would ensure that array is written with contiguous writes, which could result in a marginal performance improvement (at least for this function).

**Review by shoyer — COMMENTED** [review]
I really like the approach here, thanks for tackling this clean-up!

File: numpy/lib/shape_base.py (hunk 6 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Yep, my goal here was to leave the memory layout of the result array unaffected when compared to the old version.

In particular, the old code would always give a contiguous output, something that may have been relied upon

File: numpy/lib/shape_base.py [hunk 7]
```diff
@@ -73,66 +76,66 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
-            res = asanyarray(func1d(arr[tuple(i.tolist())], *args, **kwargs))
-            outarr[tuple(i.tolist())] = res
-            k += 1
-        if res.shape == ():
-            outarr = outarr.squeeze(axis)
-        return outarr
+
+    # arr, with the iteration axis at the end
+    dims_in = list(range(nd))
+    inarr_view = transpose(arr, dims_in[:axis] + dims_in[axis+1:] + [axis])
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Actually, only the output `transpose` does

Comment (shoyer) [inline numpy/lib/shape_base.py]: As @seberg writes in that mailing list discussion, transpose on base numpy arrays always returns a view, never a copy.

Comment (shoyer) [inline numpy/lib/shape_base.py]: But this might indeed break in surprising ways for ndarray subclasses

File: numpy/lib/shape_base.py (hunk 6 above)
Comment (shoyer) [inline numpy/lib/shape_base.py]: I think we reserve the right to change the memory layout of arrays returned by NumPy functions unless it is documented as part of the API. Users who truly rely on contiguous output should be using `np.ascontiguousarray`.

File: numpy/lib/shape_base.py (hunk 7 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: @shoyer: Would an acceptable compromise be to not call `__array_prepare__` if `tranpose` is a copy?

File: numpy/lib/shape_base.py (hunk 6 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: I guess transposing afterwards would fix the above issue

File: numpy/lib/shape_base.py (hunk 5 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: I think `MaskedConstant.__array_prepare__` was broken - fixed in #8508. I think a corollary of this is that `np.ma.apply_along_axis` can basically defer now.

File: numpy/lib/shape_base.py (hunk 6 above)
Comment (seberg) [inline numpy/lib/shape_base.py]: We have done it before (e.g. even indexing I did it). If it is obvious/possible, then "the same as before" should be preferred most of the time I would say.

Comment (shoyer) [inline numpy/lib/shape_base.py]: I agree that keeping the same memory layout should be preferred in general. But in this case, keeping the same memory layout requires jumping through some hoops, and the resulting code is much less obvious. So I would lean towards changing it.

Comment (seberg) [inline numpy/lib/shape_base.py]: Sure, I did not even check what is being checked here, just saying what my take on the policy is :).

Comment (eric-wieser) [inline numpy/lib/shape_base.py]: I'd argue that the correctness is more obvious this way around, because otherwise the inverse transposed order is needed, which is more troublesome to generate

Comment (shoyer) [inline numpy/lib/shape_base.py]: I mostly like transposing afterwards more. I don't care so much about the memory layout.

Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Actually I think you're right, it is simpler - the transpose order becomes `dims_out[:axis] + dims_out[-res.ndim:] + dims_out[axis+1:]`

Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Should the transpose happen before or after the `__array_wrap__`?

Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Ok, I'm sold on your suggestion, but for a different reason: when `res.__array_prepare__` is called, the axes of `res` (a single output) and the final output are paired as if they were broadcasted, which seems desirable for any kind of axis metadata copying

**Comment by eric-wieser** [comment]
@shoyer: Transposed moved, which now gives this `masked_array` support, leading into #8511

File: numpy/lib/shape_base.py [hunk 8]
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
+            [0, 8, 0],
+            [0, 0, 9]]])
     """
+    # handle negative axes
     arr = asanyarray(arr)
     nd = arr.ndim
     if axis < 0:
         axis += nd
-    if (axis >= nd):
+    if axis >= nd:
```
Comment (mhvk) [inline numpy/lib/shape_base.py]: while we're at it, also check that axis is not still negative?

File: numpy/lib/shape_base.py [hunk 9]
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
+    # build a buffer for storing evaluations of func1d.
+    # remove the requested axis, and add the new ones on the end.
+    # laid out so that each write is contiguous.
+    # for a tuple index inds, pre_arr[inds] = func1d(inarr_view[inds])
+    pre_shape = arr.shape[:axis] + arr.shape[axis+1:] + res.shape
+    pre_arr = zeros(pre_shape, res.dtype)
+
+    # permutation of axes such that out = pre_arr.transpose(pre_permute)
+    pre_dims = list(range(pre_arr.ndim))
+    pre_permute = pre_dims[:axis] + list(roll(pre_dims[axis:], res.ndim))
```
Comment (mhvk) [inline numpy/lib/shape_base.py]: Maybe a bit faster (no list->ndarray->list) and clearer (to me at least):
```
if res.ndim > 0:
   pre_permute = pre_dims[:axis] + pre_dims[-res.ndim:] + pre_dims[axis+red.ndim:]
```
(you'd need to rename `pre_dims` to `pre_permute` if you want the transpose at the end even for `res.ndim=0`)

File: numpy/lib/shape_base.py [hunk 10]
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
+    pre_arr[ind0] = res
+    for ind in inds:
+        pre_arr[ind] = asanyarray(func1d(inarr_view[ind], *args, **kwargs))
+
+    if not isinstance(res, matrix):
+        # wrap the array, to preserve subclasses
+        pre_arr = res.__array_wrap__(pre_arr)
+
+        # finally, rotate the inserted axes back to where they belong
+        return transpose(pre_arr, pre_permute)
```
Comment (mhvk) [inline numpy/lib/shape_base.py]: For speed for the more common case (and required with my comment above):
```
return pre_arr if res.ndim == 0 else transpose(pre_arr, pre_permute)
```

File: numpy/lib/shape_base.py [hunk 11]
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
-                n -= 1
-            i.put(indlist, ind)
-            res = asanyarray(func1d(arr[tuple(i.tolist())], *args, **kwargs))
-            outarr[tuple(i.tolist())] = res
-            k += 1
-        if res.shape == ():
-            outarr = outarr.squeeze(axis)
-        return outarr
+        # matrices have to be transposed first, because they collapse dimensions!
+        out_arr = transpose(pre_arr, pre_permute)
```
Comment (mhvk) [inline numpy/lib/shape_base.py]: same as above

File: numpy/lib/tests/test_shape_base.py [hunk 12]
```diff
@@ -28,13 +28,23 @@ def test_3d(self):
...
+        # refuses to become 1d
         def double(row):
             return row * 2
         m = np.matrix([[0, 1], [2, 3]])
+        expected = np.matrix([[0, 2], [4, 6]])
+
         result = apply_along_axis(double, 0, m)
-        assert isinstance(result, np.matrix)
+        assert_(isinstance(result, np.matrix))
+        assert_array_equal(
```
Comment (mhvk) [inline numpy/lib/tests/test_shape_base.py]: Just on one line?
```
assert_array_equal(result, expected)
```

File: numpy/lib/tests/test_shape_base.py [hunk 13]
```diff
@@ -28,13 +28,23 @@ def test_3d(self):
...
+            result, expected
+        )
+
+        result = apply_along_axis(double, 1, m)
+        assert_(isinstance(result, np.matrix))
         assert_array_equal(
-            result, np.matrix([[0, 2], [4, 6]])
+            result, expected
         )
```
Comment (mhvk) [inline numpy/lib/tests/test_shape_base.py]: same

**Comment by mhvk** [comment]
@eric-wieser - this is very nice indeed, I like the generality, and the care for subclasses. My comments are all small, but hopefully make this a little better still.

File: numpy/lib/shape_base.py [hunk 14]
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
-                ind[n] = 0
-                n -= 1
-            i.put(indlist, ind)
-            res = func1d(arr[tuple(i.tolist())], *args, **kwargs)
-            outarr[tuple(ind)] = res
-            k += 1
-        return outarr
+
+    # arr, with the iteration axis at the end
+    in_dims = list(range(nd))
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Oh, good point. `moveaxis` works with multiple axes too, so should work in both cases.

File: numpy/lib/shape_base.py [hunk 15]
```diff
@@ -73,66 +76,75 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
...
+    # invoke the function on the first item
+    ind0 = next(inds)
+    res = func1d(inarr_view[ind0], *args, **kwargs)
+    res = asanyarray(res)
+
+    # build a buffer for storing evaluations of func1d.
+    # remove the requested axis, and add the new ones on the end.
+    # laid out so that each write is contiguous.
+    # for a tuple index inds, pre_arr[inds] = func1d(inarr_view[inds])
+    pre_shape = arr.shape[:axis] + arr.shape[axis+1:] + res.shape
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Nice!

File: numpy/lib/shape_base.py (hunk 9 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Yeah, that line was inspired by a talk extolling the virtues of `std::rotate`.

Had that before, but without the if - as you correctly spot, it fails when `res.ndim == 0`. Not a huge fan of special casing that.

Either way, I reckon moveaxis will do the job here too.

**Comment by eric-wieser** [comment]
I'll do another round of clean up tomorrow to address those

**Comment by mhvk** [comment]
> I'll do another round of clean up tomorrow to address those

I think I'd go for your approach rather than `moveaxis` as it is not all that obviously handier in passing in multiple axes.

File: numpy/lib/shape_base.py (hunk 14 above)
Comment (shoyer) [inline numpy/lib/shape_base.py]: For restoring axes, you could maybe write something like:
```python
moveaxis(pre_arr, [i + nd - 1 for i in range(res.ndim)],
         [i + axis for i in range(res.ndim)])
```

I think this is a slight improvement over building up the permutation axes for transpose directly.

Comment (eric-wieser) [inline numpy/lib/shape_base.py]: I think `moveaxis` should learn a shorthand for dest such that this would work:
```
moveaxis(pre_arr, [i + nd - 1 for i in range(res.ndim)], axis)
```

File: numpy/lib/shape_base.py (hunk 8 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: I like the idea of killing this altogether, and delegating parsing `axis` to `moveaxis`

File: numpy/lib/shape_base.py [hunk 16]
```diff
@@ -103,11 +103,10 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
             % (axis, nd))
 
     # arr, with the iteration axis at the end
-    in_dims = list(range(nd))
-    inarr_view = transpose(arr, in_dims[:axis] + in_dims[axis+1:] + [axis])
+    inarr_view = moveaxis(arr, axis, -1)
```
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: To me, this seems like it would be a fair bit slower, and isn't much clearer.

File: numpy/lib/shape_base.py (hunk 9 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Or how about just
```
pre_permute = pre_dims[:axis] + pre_dims[nd-res.ndim:] + pre_dims[axis+red.ndim:]
```

Adding the `nd` there fixes the error when res.ndim= 0

File: numpy/lib/shape_base.py (hunk 8 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Alarmingly, I can't even find a precedent where numpy checks for this

File: numpy/lib/shape_base.py (hunk 16 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: Ooh, how about `np.r_[:axis, axis+1:nd, axis]`

**Comment by eric-wieser** [comment]
Keep changing my mind on this one. I think it's best left without `moveaxis`. It doesn't add much to clarity in the second case, and I think there's a benefit to having both transposes written in the same style. Also, building the axis lists directly should be a lot faster than incurring list comprehensions, then the type conversion and error checking that occurs in `moveaxis`.

`np.r_` added some minor conciseness, but at the expense of some performance

I agree that the `roll` was a little cryptic in the second case.

**Comment by mhvk** [comment]
@eric-wieser - sorry for having brought up `moveaxis` when I had also concluded that in the end it was not as good an idea -- it works well for the first case, not so well for the second.

**Comment by mhvk** [comment]
I had another look at the code and looks all OK.

**Comment by eric-wieser** [comment]
Does this need a release note?

File: numpy/lib/shape_base.py (hunk 14 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: @shoyer: Based on @mhvk's revised opinion, are you happy for this to remain as it is? Any further changes blocking this being merged?

File: numpy/lib/shape_base.py (hunk 8 above)
Comment (mhvk) [inline numpy/lib/shape_base.py]: @eric-wieser - when you add the changelog, could you also move this check up to before `axis` is possibly changed and check for too-negative as well? That way, the error message gives the actual input rather than something that's already changed. I.e.,
```
if axis < -nd or axis >= nd:
    raise ...

if axis < 0:
    axis += nd
```

**Comment by mhvk** [comment]
Always good to add in the bug-fix section!

File: numpy/lib/tests/test_shape_base.py [hunk 17]
```diff
@@ -55,9 +61,45 @@ class MinimalSubclass(np.ndarray):
             pass
         a = np.ones((6, 3)).view(MinimalSubclass)
         res = apply_along_axis(np.sum, 0, a)
-        assert isinstance(res, MinimalSubclass)
+        assert_(isinstance(res, MinimalSubclass))
         assert_array_equal(res, np.array([6, 6, 6]).view(MinimalSubclass))
 
+    def test_0d_array(self):
+        class MinimalSubclass(np.ndarray):
```
Comment (shoyer) [inline numpy/lib/tests/test_shape_base.py]: These tests for shape behavior shouldn't test subclasses at the same time. Can you move the subclass checks (if they're really necessary) to their own tests?

**Review by shoyer — COMMENTED** [review]
I have a minor suggestion for cleaning up the test suite. Otherwise, this looks good to me.

File: numpy/lib/tests/test_shape_base.py (hunk 17 above)
Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: How about `test_0d_array(self, cls=np.ndarray)`, and having the subclass tests call that with a subclass?

Comment (shoyer) [inline numpy/lib/tests/test_shape_base.py]: Sure, something like that sounds reasonable to me.

File: numpy/lib/shape_base.py (hunk 8 above)
Comment (eric-wieser) [inline numpy/lib/shape_base.py]: @mhvk: Done

File: numpy/lib/tests/test_shape_base.py (hunk 17 above)
Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: @shoyer: Done

**Comment by eric-wieser** [comment]
Ok, tests fixed, axis checking added, and release notes done too!

File: numpy/lib/tests/test_shape_base.py [hunk 18]
```diff
@@ -50,13 +56,50 @@ def minimal_function(array):
...
+        assert_(isinstance(res, cls))
+        assert_array_equal(res, np.array([6, 6, 6]).view(cls))
+
+        res = apply_along_axis(lambda x: np.squeeze(np.sum(x, keepdims=True)), 1, a)
+        assert_(isinstance(res, cls))
+        assert_array_equal(res, np.array([3, 3, 3, 3, 3, 3]).view(cls))
+
+    def test_axis_insertion(self, cls=np.ndarray):
+        a = np.arange(18).reshape((6, 3))
+        res = apply_along_axis(lambda x: np.diag(x).view(cls), 0, a)
```
Comment (shoyer) [inline numpy/lib/tests/test_shape_base.py]: Reading over these tests one last time, it occurs to me that we should be more comprehensive for testing insertion of new axes.

Right now you only test inserting at the start, but that doesn't really exercise the permutation logic in a comprehensive way. Let's also test inserting in the middle and at the end. Also, diag is axes order invariant, so none of the tests verify that new axes are inserted in the right order.

File: numpy/lib/tests/test_shape_base.py [hunk 19]
```diff
@@ -50,13 +56,50 @@ def minimal_function(array):
...
+        assert_(isinstance(res, cls))
+        assert_equal(res.ndim, 3)
+        assert_array_equal(res[:,:,0], np.diag(a[:,0]).view(cls))
```
Comment (shoyer) [inline numpy/lib/tests/test_shape_base.py]: A cleaner way to check this sort of thing is to construct the desired result array and just call `assert_array_equal` (e.g., `np.stack([np.diag(a[:,i]) for i in range(3)], axis=-1).view(cls)`). Otherwise it's easy to overlook one part of the equality check (e.g., shape, in this case).

File: numpy/lib/tests/test_shape_base.py (hunk 18 above)
Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: Maybe outer product with itself would be better than diag then?

Comment (shoyer) [inline numpy/lib/tests/test_shape_base.py]: I think the outer product is still invariant to transposes. Something like `x[:, np.newaxis] - x[np.newaxis, :]` would work, though.

File: numpy/lib/tests/test_shape_base.py (hunk 19 above)
Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: @shoyer: I was hoping I could be done with this, but you are definitely right. I'll fix those things up soon

File: numpy/lib/tests/test_shape_base.py (hunk 18 above)
Comment (shoyer) [inline numpy/lib/tests/test_shape_base.py]: On re-reading: yes, outer product with its **reverse**, would be fine

Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: You didn't miss it the first time, I did indeed suggest a plain outer product beforehand, and realized my mistake

File: numpy/lib/tests/test_shape_base.py (hunk 19 above)
Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: Ok, done.

File: numpy/lib/tests/test_shape_base.py (hunk 18 above)
Comment (eric-wieser) [inline numpy/lib/tests/test_shape_base.py]: You didn't miss it the first time, I did indeed suggest a plain outer product beforehand, and realized my mistake. This is now done too.

**Review by shoyer — APPROVED** [review]
LGTM. Will merge this shortly if no one else has further comments.

**Comment by shoyer** [comment]
@eric-wieser could you please clean up the git history a little bit? Thanks

**Comment by eric-wieser** [comment]
@shoyer: Fair, I'll give that a go over the weekend.

**Comment by eric-wieser** [comment]
@shoyer: Down from 13 commits to 8 commits. Look ok?

**Comment by shoyer** [comment]
I would usually squash down to one commit, but this is fine.
